std: refactor explanation of `NonNull`

xizheyin opened the pull request:
Fixes rust-lang/rust#141933 


I cut out the excessive explanation and used an example to explain how to maintain invariance, but I think what is quoted in the *rust reference* in the document needs to be added with a more layman's explanation and example.

(I'm not sure if I deleted too much)

r? @workingjubilee

workingjubilee commented on library/core/src/ptr/non_null.rs:
````diff
@@ -11,36 +11,23 @@ use crate::{fmt, hash, intrinsics, mem, ptr};
-/// and `LinkedList`. This is the case because they provide a public API that follows the
-/// normal shared XOR mutable rules of Rust.
+/// ```rust
+/// use std::marker::PhantomData;
+/// use std::cell::Cell;
+/// use std::ptr::NonNull;
 ///
-/// If your type cannot safely be covariant, you must ensure it contains some
-/// additional field to provide invariance. Often this field will be a [`PhantomData`]
-/// type like `PhantomData<Cell<T>>` or `PhantomData<&'a mut T>`.
+/// struct MyCell<T> {
+///     ptr: NonNull<T>,
+///     _invariant: PhantomData<Cell<T>>,
+/// }
+/// ```
````
I feel this made it too vague.

workingjubilee commented on library/core/src/ptr/non_null.rs:
```diff
@@ -11,36 +11,23 @@ use crate::{fmt, hash, intrinsics, mem, ptr};
-/// reference. When using this `From` instance without an `UnsafeCell<T>`,
-/// it is your responsibility to ensure that `as_mut` is never called, and `as_ptr`
-/// is never used for mutation.
+/// In most cases, you should use `NonNull<T>` when you want a non-null pointer. If you are unsure, using a raw pointer (`*mut T`) is also fine.
```
This isn't about variance so it probably shouldn't have been removed the same way.

xizheyin committed:
std: simplify `NonNull` variance documentation

Streamlined the explanation of covariance for `NonNull<T>`,
focusing on practical usage and reducing scary explanation.
Added a concise example for cases where invariance is required,
showing how to use `PhantomData<Cell<T>>

Signed-off-by: xizheyin <xizheyin@smail.nju.edu.cn>

workingjubilee commented:
Hm. I know I filed the issues but I wasn't really expecting to get two PRs immediately and r?'d on them, I kinda expected them to hit the rotation, and this is a slightly more tricky PR to judge when biased by being the reporter. So I'm going to pass them onto the rotation, sorry! ^^;

r? libs
